datalad/downloaders/s3.py

```diff
@@ -198,18 +212,127 @@ def _establish_session(self, url, allow_old=True):
         bucket_name = self._parse_url(url, bucket_only=True)
         if allow_old and self._bucket:
             if self._bucket.name == bucket_name:
-                lgr.debug(
-                    "S3 session: Reusing previous connection to bucket %s",
-                    bucket_name
-                )
-                return True  # we used old
+                try:
+                    self._check_credential()
+                    lgr.debug(
+                        "S3 session: Reusing previous connection to bucket %s",
+                        bucket_name
+                    )
+                    return True  # we used old
+                except AccessPermissionExpiredError:
+                    lgr.debug("S3 session: credential expired")
             else:
                 lgr.warning("No support yet for multiple buckets per S3Downloader")
 
         lgr.debug("S3 session: Reconnecting to the bucket")
-        self._bucket = self.authenticator.authenticate(bucket_name, self.credential)
+        self._bucket = try_multiple_dec_s3(self.authenticator.authenticate)(
+            bucket_name, self.credential)
         return False
 
+    def _check_credential(self):
+        """Quick check of the credential if known on either it has not expired
+
+        Raises
+        ------
+        AccessPermissionExpiredError
+          if credential is found to be expired
+        """
+        if self.credential and self.credential.is_expired:
+            raise AccessPermissionExpiredError(
+                "Credential %s has expired" % self.credential)
+
+    def _get_key(self, key_name, version_id=None, headers=None):
+        try:
+            return self._bucket.get_key(key_name, version_id=version_id, headers=headers)
+        except S3ResponseError as e:
+            if e.status != 400:
+                raise  # we will not deal with those here
+            # e.g. 400 Bad request could happen due to timed out key.
+            # Since likely things went bad if credential expired, just raise general
+            # AccessDeniedError. Logic upstream should retry
+            self._check_credential()
+            try:
+                return self._get_key_via_get(key_name, version_id=version_id, headers=headers)
```

Comment:
crap -- I forgot to add a `lgr.debug` here to actually spot such cases when they happen.  Current run 'succeeded' but I do not know if it is thanks to this code or not :-/